### PR TITLE
Use Colab as a base image.

### PR DESCRIPTION
**Pull request by djherbis**
This change makes a number of major changes:
- Colab is the base image
- uv is the main package install tool
- leveraging requirements.txt instead of many separate installs
- stop building and installing tensorflow/torch/lightbgm/jax since those are managed by the Colab base image now

In order to decide what packages to explicitly install I:
- looked at what packages are in the Colab base image
- looked at what packages were in the Kaggle image
- looked at what packages were explicitly mentioned in Kaggle Dockerfile

This may still take a few iterations to get all the right parts in the image, but this should hopefully make the image much more manageable.

http://b/365782129

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -21,66 +21,6 @@ pipeline {
...
-                --package jaxlib \
-                --version $JAX_VERSION \
-                --build-arg CUDA_MAJOR_VERSION=$CUDA_MAJOR_VERSION \
-                --build-arg CUDA_MINOR_VERSION=$CUDA_MINOR_VERSION \
-                --push
-            '''
-          }
-        }
-      }
-    }
```
Comment (djherbis) [inline Jenkinsfile:83]: These are no longer needed since they are installed in the base image.

File: test [hunk 2]
```diff
@@ -3,7 +3,7 @@ set -e
 
 IMAGE_TAG='kaggle/python-build'
 IMAGE_TAG_OVERRIDE=''
-ADDITONAL_OPTS=''
+ADDITONAL_OPTS='--runtime runc ' # Use the CPU runtime by default
```
Comment (djherbis) [inline test:6]: Colab image fails if using the nvidia runtime (which Jenkins does by default) while on a CPU VM.

File: tests/test_fastai.py [hunk 3]
```diff
@@ -27,8 +27,9 @@ def test_tabular(self):
             "/input/tests/data/train.csv",
             cont_names=["pixel"+str(i) for i in range(784)],
             y_names='label',
-            procs=[FillMissing, Categorify, Normalize])       
+            procs=[FillMissing, Categorify, Normalize])
         learn = tabular_learner(dls, layers=[200, 100])
-        learn.fit_one_cycle(n_epoch=1)
+        with learn.no_bar():
```
Comment (djherbis) [inline tests/test_fastai.py:32]: fastprogress was trying to render html while running in the background and it caused issues, so I turned off the progress bar rendering.

File: config.txt [hunk 4]
```diff
@@ -1,11 +1,2 @@
...
-CPU_BASE_IMAGE_NAME=tf2-cpu.2-16.py310
-GPU_BASE_IMAGE_NAME=tf2-gpu.2-16.py310
-LIGHTGBM_VERSION=4.2.0
-TORCH_VERSION=2.4.0
-TORCHAUDIO_VERSION=2.4.0
-TORCHVISION_VERSION=0.19.0
-JAX_VERSION=0.4.26
 CUDA_MAJOR_VERSION=12
-CUDA_MINOR_VERSION=3
+CUDA_MINOR_VERSION=2
```
Comment (rosbo) [inline config.txt:2]: I assume this downgrade is to align with Colab?

File: tests/test_lightgbm.py [hunk 5]
```diff
@@ -34,7 +34,9 @@ def test_cpu(self):
 
         self.assertEqual(1, gbm.best_iteration)
 
+    # TODO(b/381256047): Colab needs to install GPU-enabled lightgbm.
```
Comment (rosbo) [inline tests/test_lightgbm.py:37]: Do you consider this a launch blocker for this?

**Review by rosbo — APPROVED** [review]
Awesome work on this!

File: config.txt (hunk 4 above)
Comment (djherbis) [reply config.txt:2]: Yes! Colab is on 12.2, so we are too now. Sucks to go backwards, but better that we're aligned.

File: Dockerfile.tmpl [hunk 6]
```diff
@@ -1,66 +1,58 @@
...
+# Install uv & Kaggle packages
+RUN curl -LsSf https://astral.sh/uv/install.sh | sh
+RUN export PATH="${HOME}/.local/bin:${PATH}" && uv pip install --system -r /requirements.txt
+ENV PATH="~/.local/bin:${PATH}"
 
+# Install manual packages:
+# b/183041606#comment5: the Kaggle data proxy doesn't support these APIs. If the library is missing, it falls back to using a regular BigQuery query to fetch data.
+RUN uv pip uninstall --system google-cloud-bigquery-storage
+
+# NOTE(herbison): uv fails to install this for some reason
```
Comment (calderjo) [inline Dockerfile.tmpl:24]: prefer a bug over ldap, unless you intend to change it to a uv install soon.

File: kaggle_requirements.txt [hunk 7]
```diff
@@ -0,0 +1,139 @@
+altair>=5.4.0
```
Comment (calderjo) [inline kaggle_requirements.txt:1]: super optional: seem sorta alphabetically ordered, is this intentional?
do we want to add a comment at the top to enforce it? ngl it is easier to read.

File: kaggle_requirements.txt [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+tensorflow-text
+tensorflow_decision_forests
+timm
+torchinfo
+torchmetrics
+tsfresh
+vtk
+wandb
+wavio
+xgboost==2.0.3
```
Comment (calderjo) [inline kaggle_requirements.txt:137]: # b/350573866: xgboost v2.1.0 breaks learntools

File: tests/test_lightgbm.py (hunk 5 above)
Comment (djherbis) [reply tests/test_lightgbm.py:37]: re-added it in #1451

Colab is also working on adding it to the base so we can re-drop once its out in their image

**Comment by francesco-ballarin** [comment]
Hi @djherbis ,
a couple of questions:

1) where do I find the built docker image after this PR? If I pull `gcr.io/kaggle-images/python` I still get an image where `python3` comes from `conda`, which I guess is the image that was built just before this PR.
2) now that you are based on the colab image, can you consider publishing backend info (basically a dump of `apt list` and a dump of `pip freeze`) just like colab does in https://github.com/googlecolab/backend-info ?

Thanks,
Francesco

**Comment by djherbis** [comment]
> Hi @djherbis ,
> a couple of questions:
> 
> 1) where do I find the built docker image after this PR? If I pull `gcr.io/kaggle-images/python` I still get an image where `python3` comes from `conda`, which I guess is the image that was built just before this PR.
> 2) now that you are based on the colab image, can you consider publishing backend info (basically a dump of `apt list` and a dump of `pip freeze`) just like colab does in https://github.com/googlecolab/backend-info ?
> 
> Thanks,
> Francesco
> 

Hey! Our releases page at least has our pip package diff on it from image to image.
https://github.com/Kaggle/docker-python/releases

We haven't released with the colab base image yet, might happen mid December or early Jan depending on our availability for the holidays.

**Pull request merged**